node/bindnode/repr.go

```diff
@@ -0,0 +1,814 @@
+	case schema.UnionRepresentation_Stringprefix:
+		return ipld.Kind_String
+	case nil:
+		return (*_node)(w).Kind()
+	default:
+		panic(fmt.Sprintf("TODO Kind: %T", stg))
+	}
+}
+
+func outboundMappedKey(stg schema.StructRepresentation_Map, key string) string {
+	// TODO: why doesn't stg just allow us to "get" by the key string?
+	field := schema.SpawnStructField(key, "", false, false)
+	mappedKey := stg.GetFieldKey(field)
+	return mappedKey
+}
+
+func inboundMappedKey(typ *schema.TypeStruct, stg schema.StructRepresentation_Map, key string) string {
+	// TODO: can't do a "reverse" lookup... needs better API probably.
+	fields := typ.Fields()
+	for _, field := range fields {
+		mappedKey := stg.GetFieldKey(field)
+		if key == mappedKey {
+			// println(key, "rev-mapped to", field.Name())
+			return field.Name()
+		}
+	}
+	// println(key, "had no mapping")
+	return key // fallback to the same key
+}
+
+func outboundMappedType(stg schema.UnionRepresentation_Keyed, key string) string {
+	// TODO: why doesn't stg just allow us to "get" by the key string?
+	typ := schema.SpawnBool(schema.TypeName(key))
+	mappedKey := stg.GetDiscriminant(typ)
+	return mappedKey
+}
+
+func inboundMappedType(typ *schema.TypeUnion, stg schema.UnionRepresentation_Keyed, key string) string {
+	// TODO: can't do a "reverse" lookup... needs better API probably.
+	for _, member := range typ.Members() {
+		mappedKey := stg.GetDiscriminant(member)
+		if key == mappedKey {
+			// println(key, "rev-mapped to", field.Name())
+			return member.Name().String()
+		}
+	}
+	// println(key, "had no mapping")
+	return key // fallback to the same key
+}
+
+func (w *_nodeRepr) asKinded(stg schema.UnionRepresentation_Kinded, kind ipld.Kind) *_nodeRepr {
+	name := stg.GetMember(kind)
+	members := w.schemaType.(*schema.TypeUnion).Members()
+	for _, member := range members {
+		if member.Name() != name {
+			continue
+		}
+		w2 := *w
+		w2.val = w.val.FieldByName("Value").Elem()
+		w2.schemaType = member
+		return &w2
+	}
+	return nil
+}
+
+func (w *_nodeRepr) LookupByString(key string) (ipld.Node, error) {
+	if stg, ok := reprStrategy(w.schemaType).(schema.UnionRepresentation_Kinded); ok {
+		w = w.asKinded(stg, ipld.Kind_Map)
+	}
+	switch stg := reprStrategy(w.schemaType).(type) {
+	case schema.StructRepresentation_Map:
+		revKey := inboundMappedKey(w.schemaType.(*schema.TypeStruct), stg, key)
+		v, err := (*_node)(w).LookupByString(revKey)
+		if err != nil {
+			return nil, err
+		}
+		return reprNode(v), nil
+	case schema.UnionRepresentation_Keyed:
+		revKey := inboundMappedType(w.schemaType.(*schema.TypeUnion), stg, key)
+		v, err := (*_node)(w).LookupByString(revKey)
+		if err != nil {
+			return nil, err
+		}
+		return reprNode(v), nil
+	case nil:
+		v, err := (*_node)(w).LookupByString(key)
+		if err != nil {
+			return nil, err
+		}
+		return reprNode(v), nil
+	default:
+		panic(fmt.Sprintf("TODO: %T", stg))
+	}
+}
+
+func (w *_nodeRepr) LookupByNode(key ipld.Node) (ipld.Node, error) {
```

Comment:
this should work when the key is a string node, right?

Reply:
Indeed, but the same I mentioned elsewhere about tests and TODOs applies :) This PR is by no means attempting to be complete